app/controllers/mixins/breadcrumbs_mixin.rb

```diff
@@ -0,0 +1,144 @@
+module Mixins
+  module BreadcrumbsMixin
+    def self.included(c)
+      c.helper_method(:data_for_breadcrumbs)
+    end
+
+    # Main method which creates all breadcrumbs and returns them (to view page which will parse them into breadcrumbs nav)
+    def data_for_breadcrumbs
+      options = breadcrumbs_options || {}
+      options[:record_info] ||= (@record || {})
+      # diplay name, description ?
+      options[:record_title] ||= :name
+      breadcrumbs = options[:breadcrumbs] || []
+
+      # Different methods for controller with explorers and for non-explorers controllers
+
+      if @sb[:explorer].blank? && @right_cell_text.blank?
+        # Append breadcrumb from @record item (eg "Openstack")
+        breadcrumbs.push(build_breadcrumbs_no_explorer(options[:record_info], options[:record_title]))
+
+        # Append tag and policy breadcrumb if they exist
+        breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems))
+
+        # Append title breadcrumb if they exist and not same as previous breadcrumb (eg "Editing name")
+        if @title && not_show_page? && @title != breadcrumbs.compact.last.try(:[], :title)
+          breadcrumbs.push(:title => @title)
+        end
+      else
+        # Append breadcrumb from title of the accordion (eg "Policies")
+        breadcrumbs.push(:title => accord_name, :key => accord_container, :action => "accordion_select") if features?
+
+        # Append breadcrumbs created from the tree (eg "All policies > Red Hat policies > Policy 1")
+        breadcrumbs.concat(build_breadcrumbs_from_tree)
+
+        if options[:include_record] && options[:record_info].present?
+          # Append breadcrumb created from record_info if included
+          breadcrumbs.push(:title => options[:record_info][options[:record_title]])
+        else
+          # Append breadcrumb from the title of right cell
+          breadcrumbs.push(special_page_breadcrumb(@tagitems || @politems || @ownershipitems || @retireitems))
+          breadcrumbs.push(:title => @right_cell_text) if @sb["action"]
+        end
+      end
+      breadcrumbs.compact
+    end
+
+    def build_breadcrumbs_no_explorer(record_info, record_title)
+      if record_info[record_title]
+        notshow = restful? ? "" : "show"
+        breadcrumb_url = File.join(controller_url, notshow, record_info[:id].to_s)
```

Comment:
Don't use `File.join` for urls, the escaping rules are completely diferent, it doesn't make sense.

Reply:
`"#{ ...` was used before, but then I was told not to use it (by @skateman ) and use URI.join instead... but it didn't work properly (I think it wanted to get base_url), but file.join works well as it just adds `/` between parameters

Reply:
Of course :) I agree that string interpolation is subotimal, but `File.join` is misleading. And `URI::join` is just bad, yeah.

The best thing would be to always use `url_for` and not have to deal with urls directly, but that's probably not realistic because of the restful mess. 

What about something like:

```ruby
def url(*components)
  File.join('/', *(components.map(&.to_s)))
end
```

That way, we can write `url(controller_name, notshow, record_info[:id])` which at least looks semantically right, and we don't have to deal with explicit `.to_s` and having to make the first arg an absolute path.